### PR TITLE
Merge dubbo-serialization-avro into incubator-dubbo #3708

### PR DESCRIPTION
**Pull request by CrazyHZM**
## What is the purpose of the change
Merge dubbo-serialization-avro into incubator-dubbo #3708

## Brief changelog

XXXXX

## Verifying this change

XXXXX

Follow this checklist to help us incorporate your contribution quickly and easily:

- [x] Make sure there is a [GITHUB_issue](https://github.com/apache/incubator-dubbo/issues) field for the change (usually before you start working on it). Trivial changes like typos do not require a GITHUB issue. Your pull request should address just this issue, without pulling in other changes - one PR resolves one issue.
- [ ] Format the pull request title like `[Dubbo-XXX] Fix UnknownException when host config not exist #XXX`. Each commit in the pull request should have a meaningful subject line and body.
- [ ] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [ ] Write necessary unit-test to verify your logic correction, more mock a little better when cross module dependency exist. If the new feature or significant change is committed, please remember to add integration-test in [test module](https://github.com/apache/incubator-dubbo/tree/master/dubbo-test).
- [ ] Run `mvn clean install -DskipTests=false` & `mvn clean test-compile failsafe:integration-test` to make sure unit-test and integration-test pass.
- [ ] If this contribution is large, please follow the [Software Donation Guide](https://github.com/apache/incubator-dubbo/wiki/Software-donation-guide).

### CODE REVIEW

**Comment by CrazyHZM** [comment]
@ralf0131 
Can you help me see where the license is wrong?

**Comment by tswstarplanet** [comment]
You can delete the line "<?xml version="1.0" encoding="UTF-8"?>"

**Comment by CrazyHZM** [comment]
> You can delete the line "xml version="1.0" encoding="UTF-8""


I tried to do this, but I can't solve it.

**Comment by ralf0131** [comment]
You can run the UT locally and check the output to see what is the problem. For example: /home/travis/build/apache/incubator-dubbo/dubbo-serialization/dubbo-serialization-avro/target/rat.txt

File: dubbo-serialization/dubbo-serialization-avro/pom.xml [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+  See the License for the specific language governing permissions and
+  limitations under the License.
+  -->
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <modelVersion>4.0.0</modelVersion>
+    <parent>
+        <groupId>org.apache.dubbo</groupId>
+        <artifactId>dubbo-serialization</artifactId>
+        <version>2.7.1-SNAPSHOT</version>
```
Comment (ralf0131) [inline dubbo-serialization/dubbo-serialization-avro/pom.xml]: Hi, since 2.7.1 has been released, could you update it to 2.7.2-SNAPSHOT?

**Comment by ralf0131** [comment]
Hi, 

While I am reviewing this pull request, I feel more people should come in and take a look. Please request @beiwei30 to review as well.

File: dubbo-serialization/dubbo-serialization-avro/src/main/java/org/apache/dubbo/common/serialize/avro/AvroObjectInput.java [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+import org.apache.dubbo.common.serialize.ObjectInput;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.lang.reflect.Type;
+import java.util.HashMap;
+import java.util.Map;
+
+public class AvroObjectInput implements ObjectInput{
+	private DecoderFactory decoderFactory=DecoderFactory.get();;
```
Comment (kexianjun) [inline dubbo-serialization/dubbo-serialization-avro/src/main/java/org/apache/dubbo/common/serialize/avro/AvroObjectInput.java]: can we make decoderFactory as a static field?

File: dubbo-serialization/dubbo-serialization-avro/src/main/java/org/apache/dubbo/common/serialize/avro/AvroObjectOutput.java [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+import org.apache.avro.reflect.ReflectDatumWriter;
+import org.apache.avro.util.Utf8;
+import org.apache.dubbo.common.serialize.ObjectOutput;
+
+import java.io.IOException;
+import java.io.OutputStream;
+import java.util.Arrays;
+
+public class AvroObjectOutput implements ObjectOutput {
+    private EncoderFactory encoderFactory = EncoderFactory.get();
```
Comment (kexianjun) [inline dubbo-serialization/dubbo-serialization-avro/src/main/java/org/apache/dubbo/common/serialize/avro/AvroObjectOutput.java]: encoderFactory also as above

File: dubbo-serialization/dubbo-serialization-avro/pom.xml (hunk 1 above)
Comment (CrazyHZM) [reply dubbo-serialization/dubbo-serialization-avro/pom.xml]: completed.

File: dubbo-serialization/dubbo-serialization-api/pom.xml [hunk 4]
```diff
@@ -55,5 +55,9 @@ limitations under the License.
             <groupId>de.ruedigermoeller</groupId>
             <artifactId>fst</artifactId>
         </dependency>
+        <dependency>
```
Comment (ralf0131) [inline dubbo-serialization/dubbo-serialization-api/pom.xml]: There is no need to depend avro here.

File: dubbo-common/pom.xml [hunk 5]
```diff
@@ -77,5 +77,9 @@
             <groupId>de.ruedigermoeller</groupId>
             <artifactId>fst</artifactId>
         </dependency>
+        <dependency>
```
Comment (ralf0131) [inline dubbo-common/pom.xml]: No need to add it here.

File: dubbo-serialization/dubbo-serialization-api/pom.xml (hunk 4 above)
Comment (CrazyHZM) [reply dubbo-serialization/dubbo-serialization-api/pom.xml]: Fixed.

File: dubbo-common/pom.xml (hunk 5 above)
Comment (CrazyHZM) [reply dubbo-common/pom.xml]: Fixed.

File: dubbo-serialization/dubbo-serialization-avro/src/main/java/org/apache/dubbo/common/serialize/avro/AvroObjectInput.java (hunk 2 above)
Comment (ralf0131) [reply dubbo-serialization/dubbo-serialization-avro/src/main/java/org/apache/dubbo/common/serialize/avro/AvroObjectInput.java]: It is fixed.

File: dubbo-serialization/dubbo-serialization-avro/src/main/java/org/apache/dubbo/common/serialize/avro/AvroObjectOutput.java (hunk 3 above)
Comment (ralf0131) [reply dubbo-serialization/dubbo-serialization-avro/src/main/java/org/apache/dubbo/common/serialize/avro/AvroObjectOutput.java]: Fixed.

**Pull request merged**